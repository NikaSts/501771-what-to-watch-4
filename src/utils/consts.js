export const Tab = {
  OVERVIEW: `Overview`,
  REVIEWS: `Reviews`,
  DETAILS: `Details`,
};

export const tabs = [Tab.OVERVIEW, Tab.REVIEWS, Tab.DETAILS];

export const DEFAULT_GENRE = `All genres`;

export const MOVIES_TO_SHOW_COUNT = 8;
export const MAX_GENRES_TO_SHOW = 10;
export const MAX_SIMILAR_MOVIES = 4;
export const MAX_ACTORS_TO_SHOW = 4;

export const Rating = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

export const Score = {
  BAD: 60,
  NORMAL: 70,
  GOOD: 80,
  VERY_GOOD: 90,
  AWESOME: 100,
};

export const DELAY = 1000;

export const ActionType = {
  OPEN_FULL_SCREEN_PLAYER: `openFullScreenPlayer`,
  CLOSE_FULL_SCREEN_PLAYER: `closeFullScreenPlayer`,
  GET_MOVIES: `getMovies`,
  GET_PROMO_MOVIE: `getPromoMovie`,
  GET_REVIEWS: `getReviews`,
  REQUIRE_AUTHORIZATION: `requireAuthorization`,
  IS_AUTHORIZING: `isAuthorizing`,
  IS_NOT_AUTHORIZING: `isNotAuthorizing`,
  IS_AUTHORIZATION_ERROR: `isAuthorizationError`,
  LOAD_USER_DATA: `loadUserData`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const EntryPoint = {
  REVIEWS: `/wtw/comments`,
  FAVORITE: `/wtw/favorite`,
  MOVIES: `/wtw/films`,
  PROMO: `/wtw/films/promo`,
  LOGIN: `/wtw/login`,
};

export const TIMER = 5000;

export const Error = {
  UNAUTHORIZED: 401,
};

export const URL = `https://4.react.pages.academy`;

export const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  MOVIE_PAGE: `/films`,
};

export const TEXTAREA_COLOR = `rgba(255, 255, 255, 0.3)`;
