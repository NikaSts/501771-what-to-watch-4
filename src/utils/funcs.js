import {DEFAULT_GENRE, Rating, Score} from './consts';

export const getRandomNumber = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const getRandomItem = (items) => {
  return items[getRandomNumber(0, items.length)];
};

export const getUrlFromTitle = (title) => {
  return title
    .toLowerCase()
    .split(` `)
    .join(`-`);
};

export const getRandomBoolean = () => {
  return Math.random() > 0.5;
};

export const getRandomItems = (items) => {
  const result = [];
  [...items].forEach((item) => {
    if (getRandomBoolean()) {
      result.push(item);
    }
  });
  return result;
};

export const suffleItems = (items) => {
  for (let i = items.length - 1; i > 0; i--) {
    const randomIndex = Math.floor(Math.random() * (i + 1));
    const temp = items[randomIndex];
    items[randomIndex] = items[i];
    items[i] = temp;
  }
  return items;
};

export const getRatingLevel = (score) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const getMoviesToShow = (movies, activeGenre) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};
