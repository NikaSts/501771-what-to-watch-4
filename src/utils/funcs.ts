
import {DEFAULT_GENRE, Rating, Score} from './consts';
import {MovieType} from '../types';

interface ExtendInterface {
    <T>(state: T, newStateValue: T): T;
}

export const extend: ExtendInterface = (state, newStateValue) => {
  return Object.assign({}, state, newStateValue);
};

export const getRatingLevel = (score: number) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const filterMovies = (movies: Array<MovieType>, activeGenre: string = DEFAULT_GENRE) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};

export const getGenresFromMovies = (movies: Array<MovieType>) => {
  return [DEFAULT_GENRE, ...new Set(movies.map((movie) => movie.genre))];
};

export const getMovieById = (movies: Array<MovieType>, id: number) => {
  return movies.find((movie) => movie.id === id);
};

export const formatTimeWithSeconds = (time: number) => {
  const minutes = Math.trunc(time / 60);
  const seconds = Math.trunc(time % 60);
  const hours = Math.trunc(minutes / 60);

  return [
    (`0` + hours).slice(-2),
    (`0` + minutes).slice(-2),
    (`0` + seconds).slice(-2)
  ].join(`:`);
};

export const formatTime = (time: number) => {
  const hours = Math.trunc(time / 60);
  const minutes = time % 60;

  if (hours) {
    return `${hours}h ${minutes}m`;
  }
  return `${minutes}m`;
};
