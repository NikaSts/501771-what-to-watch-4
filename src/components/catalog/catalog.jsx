import React from 'react';
import {connect} from 'react-redux';
import cn from 'classnames';
import MovieList from '../movie-list/movie-list';
import GenreList from '../genre-list/genre-list';
import withShowMoreButton from '../../hocs/with-show-more-button/with-show-more-button';
import {func, arrayOf, string, bool} from 'prop-types';
import {movieType} from '../../types';
import {changeActiveMovie} from '../../store/actions';
import {getGenres, filterMovies} from '../../utils/funcs';
import {DEFAULT_GENRE, MAX_SIMILAR_MOVIES} from '../../utils/consts';

const WrappedMovieList = withShowMoreButton(MovieList);

const Catalog = (props) => {
  const {
    isMain, movies, onMovieTitleClick, activeItem, onItemClick,
  } = props;
  const genres = getGenres(movies);
  const moviesByGenre = filterMovies(movies, activeItem);
  const moviesToShow = [...moviesByGenre].splice(0, MAX_SIMILAR_MOVIES);
  const catalogClass = cn({
    'catalog': true,
    'catalog--like-this': !isMain,
  });
  const catalogTitleClass = cn({
    'catalog__title': true,
    'visually-hidden': isMain,
  });

  return (
    <section className={catalogClass}>
      <h2 className={catalogTitleClass}>
        {isMain ? `Catalog` : `More like this`}
      </h2>
      {isMain && <GenreList
        genres={genres}
        activeGenre={activeItem}
        onGenreClick={onItemClick}
      />}
      {isMain
        ? <WrappedMovieList
          movies={moviesByGenre}
          onMovieTitleClick={onMovieTitleClick}
        />
        : <MovieList
          movies={moviesToShow}
          onMovieTitleClick={onMovieTitleClick}
        />}
    </section>
  );
};


Catalog.defaultProps = {
  isMain: false,
  activeItem: DEFAULT_GENRE,
};

Catalog.propTypes = {
  isMain: bool.isRequired,
  movies: arrayOf(movieType.isRequired).isRequired,
  onMovieTitleClick: func.isRequired,
  activeItem: string,
  onItemClick: func,
};

const mapStateToProps = ({movies}) => ({movies});

const mapDispatchToProps = (dispatch) => ({
  onMovieTitleClick(activeMovie) {
    dispatch(changeActiveMovie(activeMovie));
  }
});


export {Catalog};
export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
