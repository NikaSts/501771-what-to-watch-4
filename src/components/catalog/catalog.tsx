import * as React from 'react';
import {connect} from 'react-redux';
import cn from 'classnames';

import CatalogList from '../catalog-list/catalog-list';
import CatalogFilter from '../catalog-filter/catalog-filter';
import withShowMoreButton from '../../hocs/with-show-more-button/with-show-more-button';
import {MovieType} from '../../types';
import {getMovies, getGenres} from '../../store/movies/selectors';
import {filterMovies} from '../../utils/funcs';
import {DEFAULT_GENRE, MAX_SIMILAR_MOVIES, Page} from '../../utils/consts';

const CatalogListWithShowMoreButton = withShowMoreButton(CatalogList);

interface CatalogProps {
  currentPage: string;
  movies?: Array<MovieType>;
  favoriteMovies?: Array<MovieType>;
  genres?: Array<string>;
  activeItem?: string;
  onItemClick?: () => void;
}

const Catalog: React.FC<CatalogProps> = ({
  currentPage, movies, favoriteMovies, genres, activeItem, onItemClick,
}: CatalogProps) => {
  const isMainPage: boolean = currentPage === Page.MAIN;
  const isMoviePage: boolean = currentPage === Page.MOVIE_PAGE;
  const isMyListPage: boolean = currentPage === Page.MY_LIST;

  const moviesByGenre: Array<MovieType> = isMainPage && filterMovies(movies, activeItem);
  const similarMovies: Array<MovieType> = isMoviePage && filterMovies(movies, activeItem)
    .splice(0, MAX_SIMILAR_MOVIES);
  const hasFavoriteMovies: boolean = isMyListPage && favoriteMovies.length > 0;

  const catalogClass = cn({
    'catalog': true,
    'catalog--like-this': isMoviePage,
  });
  const catalogTitleClass = cn({
    'catalog__title': true,
    'visually-hidden': isMainPage || isMyListPage,
  });

  return (
    <section className={catalogClass}>
      <h2 className={catalogTitleClass}>
        {isMainPage || isMyListPage ? `Catalog` : `More like this`}
      </h2>

      {isMainPage && <React.Fragment>
        <CatalogFilter
          genres={genres}
          activeGenre={activeItem}
          onGenreClick={onItemClick}
        />
        <CatalogListWithShowMoreButton
          movies={moviesByGenre}
        />
      </React.Fragment> }

      {isMoviePage && <CatalogList
        movies={similarMovies}
      />}

      {(isMyListPage && hasFavoriteMovies) && <CatalogList
        movies={favoriteMovies}
      />}
    </section>
  );
};

Catalog.defaultProps = {
  activeItem: DEFAULT_GENRE,
};

const mapStateToProps = (state) => ({
  movies: getMovies(state),
  genres: getGenres(state),
});


export {Catalog};
export default connect(mapStateToProps)(Catalog);
