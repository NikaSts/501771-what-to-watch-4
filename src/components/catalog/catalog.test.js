import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {Router} from 'react-router-dom';
import history from '../../history';
import {Catalog} from './catalog';
import {Page} from '../../utils/consts';

const mockStore = configureStore([]);

const movies = [
  {
    id: 3,
    title: `First Movie`,
    runTime: 1,
    genre: ``,
    releaseDate: 1,
    poster: ``,
    previewImage: ``,
    backgroundImage: ``,
    backgroundColor: ``,
    ratingScore: 1,
    ratingLevel: ``,
    ratingCount: 1,
    description: ``,
    director: ``,
    stars: [``],
    preview: ``,
    video: ``,
    isFavorite: false,
  },
  {
    id: 4,
    title: `Second Movie`,
    runTime: 1,
    genre: ``,
    releaseDate: 1,
    poster: ``,
    previewImage: ``,
    backgroundImage: ``,
    backgroundColor: ``,
    ratingScore: 1,
    ratingLevel: ``,
    ratingCount: 1,
    description: ``,
    director: ``,
    stars: [``],
    preview: ``,
    video: ``,
    isFavorite: false,
  },
];
const genres = [`a`, `b`];
const favoriteMovies = [{
  id: 1,
  title: `Favorite Movie`,
  runTime: 1,
  genre: ``,
  releaseDate: 1,
  poster: ``,
  previewImage: ``,
  backgroundImage: ``,
  backgroundColor: ``,
  ratingScore: 1,
  ratingLevel: ``,
  ratingCount: 1,
  description: ``,
  director: ``,
  stars: [``],
  preview: ``,
  video: ``,
  isFavorite: true,
}];
const noop = () => { };

it(`Catalog should render correctly`, () => {
  const store = mockStore({
    MOVIES: {
      movies,
      favoriteMovies,
    },
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <Catalog
              currentPage={Page.MAIN}
              movies={movies}
              genres={genres}
              favoriteMovies={favoriteMovies}
              onItemClick={noop}
            />
          </Router>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        })
      .toJSON();

  expect(tree).toMatchSnapshot();
});
