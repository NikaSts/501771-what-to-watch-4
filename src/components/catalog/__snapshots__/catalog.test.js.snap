// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog should render correctly 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        a
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        b
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        href="/films/3"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height="175"
            preload="none"
            width="280"
          />
        </div>
      </a>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/3"
          onClick={[Function]}
        >
          First Movie
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        href="/films/4"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height="175"
            preload="none"
            width="280"
          />
        </div>
      </a>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/4"
          onClick={[Function]}
        >
          Second Movie
        </a>
      </h3>
    </article>
  </div>
</section>
`;
